lib/pbench/server/cache_manager.py

```diff
@@ -1393,16 +1530,35 @@ def find_dataset(self, dataset_id: str) -> Tarball:
         if dataset_id in self.datasets:
             return self.datasets[dataset_id]
 
-        # The dataset isn't already known; so search for it in the ARCHIVE tree
-        # and (if found) discover the controller containing that dataset.
-        for dir_entry in self.archive_root.iterdir():
-            if dir_entry.is_dir() and dir_entry.name != self.TEMPORARY:
-                for file in dir_entry.glob(f"**/*{Dataset.TARBALL_SUFFIX}"):
-                    md5 = get_tarball_md5(file)
-                    if md5 == dataset_id:
-                        self._add_controller(dir_entry)
-                        return self.datasets[dataset_id]
-        raise TarballNotFound(dataset_id)
+        # The dataset isn't already known; so follow the tarball-path to build
+        # just the necessary controller and tarball objects.
+        start = time.time()
+        sql = start
+        try:
+            dataset = Dataset.query(resource_id=dataset_id)
+            sql = time.time()
+            tarball = Path(Metadata.getvalue(dataset, Metadata.TARBALL_PATH))
+            controller_dir = tarball.parent
+            if controller_dir.name == dataset_id:
+                controller_dir = controller_dir.parent
+        except Exception as e:
+            raise TarballNotFound(dataset_id) from e
```

Comment:
Which of these statements is likely to raise an exception?  (The first one, obviously...but, are there any others?  Should some or all of the remainder be moved out of the `try` block?)